src/dynamic_fmt.c

```diff
@@ -404,6 +408,46 @@ char *RemoveHEX(char *output, char *input)
 	return output;
 }
 
+// Add $HEX$ encoding for salts where needed.  For canonicalized pot storage, we encode anything
+// containing unprintables or field separators, but never anything else
+static char *AddHEX(char *ciphertext) {
+	static char out[1024];
+	char *cp;
+
+	if (strncmp(ciphertext, "$dynamic_", 9) && strncmp(ciphertext, "@dynamic=", 9))
+		return ciphertext;  // not a dynamic format, so we can not 'fix' it.
+
+	// ok, see if this is salted:
+	if (!(cp = strchr(&ciphertext[1], ciphertext[0])) || !(cp = strchr(&cp[1], '$')))
+		return ciphertext;  // not a salted format.
+
+	if (!strncmp(cp, "$HEX$", 5))
+		return ciphertext;  // already HEX$
+
+	char *s = cp++;
```

Comment:
Maybe `unsigned char *` and cast, so that we don't rely on signed `*s < ' '` also catching 8-bit (although it does).

Reply:
I'm not sure why you want this but I'm all ears

Reply:
A matter of style. If we do rely on the signed compare catching 8-bit elsewhere, then no reason to change the code here - OK to leave it as you have it.

Reply:
Hmm... so as currently written, the `<` catches nearly all of the unprintables while the `>` catch only single char or so, right? So it works but not quite as I pictured it. I agree it might be a good habit to use unsigned then.

Reply:
> as currently written, the `<` catches nearly all of the unprintables while the `>` catch only single char or so, right?

Yes, that's how it works with signed `char`. BTW, the signedness of `char` is implementation specific, so as written the checks works differently under the hood on different platforms. Luckily, the outcome is the same.